File: Frontend/trackr-administrador/src/app/shared/http/catalogo/tipo-compania.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { TipoCompania } from '@models/catalogo/tipo-compania';

@Injectable()
export class TipoCompaniaService {
  private dataUrl = 'tipoCompania/';

  constructor(public http: HttpClient) {}

  public consultarParaSelector(): Observable<TipoCompania[]> {
    return this.http.get<TipoCompania[]>(this.dataUrl + 'consultarParaSelector');
  }

  public consultar(idTipoCompania: number): Observable<TipoCompania> {
    return this.http.get<TipoCompania>(this.dataUrl + `consultar/${idTipoCompania}`);
  }

  public consultarTodosParaGrid(): Observable<TipoCompania[]> {
    return this.http.get<TipoCompania[]>(this.dataUrl + `consultarTodosParaGrid`);
  }

  public consultarPorClave(claveTipoCompania: string): Observable<TipoCompania> {
    return this.http.get<TipoCompania>(this.dataUrl + `consultarPorClave/${claveTipoCompania}`);
  }

  public agregar(tipoCompania: TipoCompania): Observable<void> {
    return this.http.post<void>(this.dataUrl + `agregar`, tipoCompania);
  }

  public editar(tipoCompania: TipoCompania): Observable<void> {
    return this.http.put<void>(this.dataUrl + `editar`, tipoCompania);
  }
  public eliminar(idTipoCompania: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idTipoCompania}`);
  }

}

File: Backend/TrackrAPI/TrackrAPI/Services/Catalogo/TipoCompaniaService.cs
using TrackrAPI.Dtos.Catalogo;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Catalogo;
using System.Collections.Generic;
using System.Transactions;

namespace TrackrAPI.Services.Catalogo
{
    public class TipoCompaniaService
    {
        private ITipoCompaniaRepository tipoCompaniaRepository;
        private TipoCompaniaValidatorService tipoCompaniaValidatorService;

        public TipoCompaniaService(ITipoCompaniaRepository tipoCompaniaRepository, TipoCompaniaValidatorService tipoCompaniaValidatorService)
        {
            this.tipoCompaniaRepository = tipoCompaniaRepository;
            this.tipoCompaniaValidatorService = tipoCompaniaValidatorService;
        }

        public TipoCompania Consultar(int idTipoCompaia)
        {
            return tipoCompaniaRepository.Consultar(idTipoCompaia);
        }

        public IEnumerable<TipoCompaniaSelectorDto> ConsultarParaSelector()
        {
            return tipoCompaniaRepository.ConsultarParaSelector();
        }

        public IEnumerable<TipoCompania> ConsultarTodosParaGrid()
        {
            return tipoCompaniaRepository.ConsultarTodosParaGrid();
        }

        public TipoCompania ConsultarPorClave(string claveTipoCompania)
        {
            return tipoCompaniaRepository.ConsultarPorClave(claveTipoCompania);
        }

        public int Agregar(TipoCompania tipoCompania)
        {
            tipoCompaniaValidatorService.ValidarAgregar(tipoCompania);
            tipoCompaniaRepository.Agregar(tipoCompania);
            return tipoCompania.IdTipoCompania;
        }

        public void Editar(TipoCompania tipoCompania)
        {
            tipoCompaniaValidatorService.ValidarEditar(tipoCompania);
            tipoCompaniaRepository.Editar(tipoCompania);
        }

        public void Eliminar(int idTipoCompania)
        {
            using (TransactionScope scope = new TransactionScope())
            {
                var tipoCompania = tipoCompaniaRepository.Consultar(idTipoCompania);
                tipoCompaniaValidatorService.ValidarEliminar(idTipoCompania);
                tipoCompaniaRepository.Eliminar(tipoCompania);
                scope.Complete();
            }
        }

    }
}
